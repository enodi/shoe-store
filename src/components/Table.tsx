import React, {useState, useMemo, useEffect} from "react";
import {
  Table,
  Thead,
  Tbody,
  Tr,
  Th,
  Td,
  Box,
  Heading
} from "@chakra-ui/react"
import ReactPaginate from "react-paginate";
import {StoreType} from "../helper/types";
import "../style.css"

interface Props {
  tableHeader: string[];
  data?: StoreType[];
  label?: string;
  showPagination?: boolean;
}

const TableComponent: React.FC<Props> = ({data, tableHeader, label, showPagination}) => {
  const [currentItems, setCurrentItems] = useState<StoreType[]>([]);
  const [pageCount, setPageCount] = useState(0);
  const [itemOffset, setItemOffset] = useState(0);
  const itemsPerPage = 10;
  const result = useMemo(() => {
    return data && data.length > 0 ? data : [];
  }, [data]);

  useEffect(() => {
    const endOffset = itemOffset + itemsPerPage;
    setCurrentItems(result.slice(itemOffset, endOffset));
    setPageCount(Math.ceil(result.length / itemsPerPage));
  }, [itemOffset, itemsPerPage, result]);

  const handlePageClick = (event: any) => {
    const newOffset = (event.selected * itemsPerPage) % result.length;
    setItemOffset(newOffset);
  };

  return (
    <Box mb={20}>
      {label && <Heading sx={{color:"#272A30", fontSize: "24px", fontWeight: 500, mb: "10px"}}>{label}</Heading>}
      <Table variant="simple">
        <Thead sx={{ bg: "gray.400", color: "#FFF" }}>
          <Tr>
            {tableHeader.map((header, index) => <Th key={index} sx={{fontSize: 16, color: '#FFF'}}>{header}</Th>)}
          </Tr>
        </Thead>
        <Tbody>
          {currentItems && currentItems.length > 0 ? currentItems.map((item: StoreType, index: number) => (
            <Tr key={index}>
              <Td>{item.store}</Td>
              <Td>{item.model}</Td>
              <Td isNumeric>{item.inventory}</Td>
            </Tr>
          ))
        : <Tr><Td colSpan={3} textAlign="center">No data available</Td></Tr>}
        </Tbody>
      </Table>
      {showPagination && currentItems.length > 0 && (
        <Box mt={10} sx={{display: "flex", justifyContent: "flex-end"}}>
          <ReactPaginate
            breakLabel="..."
            nextLabel="next >"
            onPageChange={handlePageClick}
            pageRangeDisplayed={5}
            pageCount={pageCount}
            previousLabel="< prev"
            marginPagesDisplayed={1}
            containerClassName="pagination"
            pageClassName="page-item"
            activeClassName="active"
            previousClassName="page-item"
            nextClassName="page-item"
          />
        </Box> 
      )}
    </Box>
  );
}

export default TableComponent;
