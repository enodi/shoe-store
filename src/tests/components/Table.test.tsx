import { shallow } from "enzyme";
import Table from "../../components/Table";

describe("Table Component", () => {
  let wrapper: any;
  beforeEach(() => {
      wrapper = shallow(<Table
        tableHeader={['Store', 'Shoe model', 'Quantity left']}
        data={[]}
        label="Latest Sales Completed"
      />);
  });

  it("should render Table component correctly", () => {
      expect(wrapper).toMatchSnapshot();
  });

  it("should contain Heading element with text", () => {
    expect(wrapper.find("Heading").text()).toEqual("Latest Sales Completed");
});
});