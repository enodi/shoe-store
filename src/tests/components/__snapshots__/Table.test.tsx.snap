// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component should render Table component correctly 1`] = `
<Box
  mb={20}
>
  <Heading
    sx={
      Object {
        "color": "#272A30",
        "fontSize": "24px",
        "fontWeight": 500,
        "mb": "10px",
      }
    }
  >
    Latest Sales Completed
  </Heading>
  <Table
    variant="simple"
  >
    <ForwardRef
      sx={
        Object {
          "bg": "gray.400",
          "color": "#FFF",
        }
      }
    >
      <ForwardRef>
        <ForwardRef
          key="0"
          sx={
            Object {
              "color": "#FFF",
              "fontSize": 16,
            }
          }
        >
          Store
        </ForwardRef>
        <ForwardRef
          key="1"
          sx={
            Object {
              "color": "#FFF",
              "fontSize": 16,
            }
          }
        >
          Shoe model
        </ForwardRef>
        <ForwardRef
          key="2"
          sx={
            Object {
              "color": "#FFF",
              "fontSize": 16,
            }
          }
        >
          Quantity left
        </ForwardRef>
      </ForwardRef>
    </ForwardRef>
    <ForwardRef>
      <ForwardRef>
        <ForwardRef
          colSpan={3}
          textAlign="center"
        >
          No data available
        </ForwardRef>
      </ForwardRef>
    </ForwardRef>
  </Table>
</Box>
`;
